krabs/krabs/etw.hpp

```diff
@@ -185,6 +197,25 @@ namespace krabs { namespace details {
         return query_trace();
     }
 
+    template <typename T>
+    void trace_manager<T>::set_trace_information(
+        TRACE_INFO_CLASS informationClass,
+        PVOID traceInformation,
+        ULONG informationLength)
+    {
+        if (trace_.registrationHandle_ == INVALID_PROCESSTRACE_HANDLE) {
```

Comment:
is this check and call needed? my preference would be to wrap TraceSetInformation as closely as possible without doing extra work on behalf of the caller, like calling `register_trace` if they haven't yet

Reply:
'set_trace_information` needs the session handle, which is currently assigned by the blocking call to 'trace::start()'. This is impractical as we call 'trace::start()' in its own thread. We would then have to somehow synchronize with the assignment of the session handle before safely calling 'set_trace_information'. At the root, the fact 'trace::start' is responsible for both calls to 'StartTrace' and 'OpenTrace' is the issue here. The proposed change allows to split those calls if needed by calling `StartTrace`, which assigns the session handle, before the blocking call to 'trace::start()'. Am I missing something?

Reply:
Got it! Rather than adding a call to `register_trace` here, would this sequence of calls give you the outcome you're looking for?

```
krabs::trace trace;
krabs::guid id(L"{A0C1853B-5C40-4B15-8766-3CF1C58F985A}");
provider<> powershell(id);
trace.enable(powershell);
trace.open(); // register_trace is called in here
trace.set_trace_information(...); // new
trace.process(); // this blocks
```

Reply:
Ah yes, I was indeed missing something :-). Thanks, yep this is perfect, I verified that it works. I removed the checks and modified the function comment for `set_trace_information`.